parser/src/infix.rs

```diff
@@ -0,0 +1,305 @@
+///
+/// ```text
+/// Arg:  1
+/// Op:   +
+/// Arg:  2
+/// Op:   ^
+/// Arg:  4
+/// Op:   *
+/// Arg:  6
+/// Op:   -
+/// Arg:  8
+/// ```
+struct Infixes<'a, Id: 'a> {
+    /// The next part of the expression that we need to flatten
+    remaining_expr: Option<Box<SpannedExpr<Id>>>,
+    /// Cached operator from a previous iteration
+    next_op: Option<TypedIdent<Id>>,
+    symbols: &'a IdentEnv<Ident = Id>,
+}
+
+impl<'a, Id: 'a + Clone> Infixes<'a, Id> {
+    fn new(expr: SpannedExpr<Id>, symbols: &'a IdentEnv<Ident = Id>) -> Infixes<'a, Id> {
+        Infixes {
+            remaining_expr: Some(Box::new(expr)),
+            next_op: None,
+            symbols: symbols,
+        }
+    }
+
+    fn get_precedence(&self, op: &TypedIdent<Id>) -> i32 {
+        get_precedence(self.symbols.string(&op.name))
+    }
+
+    fn get_fixity(&self, op: &TypedIdent<Id>) -> Fixity {
+        get_fixity(self.symbols.string(&op.name))
+    }
+
+    /// Reconstruct the infix expression using the correct associativities
+    /// and precedences.
+    ///
+    /// Inspired by [`Language.Haskell.Infix`].
+    ///
+    /// [`Language.Haskell.Infix`]: https://hackage.haskell.org/package/infix-0.1.1/docs/src/Language-Haskell-Infix.html
+    fn reparse(mut self) -> SpannedExpr<Id> {
+        use base::pos;
+
+        let mut arg_stack = Vec::new();
+        let mut op_stack = Vec::new();
+
+        let make_op = |lhs: Box<SpannedExpr<Id>>, op, rhs: Box<SpannedExpr<Id>>| {
+            let span = pos::span(lhs.span.start, rhs.span.end);
+            Box::new(pos::spanned(span, Expr::Infix(lhs, op, rhs)))
+        };
+
+        loop {
+            if arg_stack.len() == 2 && op_stack.len() == 1 {
+                let lhs = arg_stack.pop().unwrap();
+                let rhs = arg_stack.pop().unwrap();
+                let op = op_stack.pop().unwrap();
+
+                arg_stack.push(make_op(lhs, op, rhs));
+            } else {
+                use std::cmp::Ordering::*;
+                use self::InfixExpr::*;
+
+                match self.next() {
+                    None => break,
+                    Some(Arg(next_expr)) => arg_stack.push(next_expr),
+                    Some(Op(next_op)) => {
+                        if op_stack.is_empty() {
+                            op_stack.push(next_op);
+                            continue;
+                        }
+
+                        let stack_op = op_stack.pop().unwrap();
+
+                        match i32::cmp(&self.get_precedence(&next_op),
+                                       &self.get_precedence(&stack_op)) {
+                            Greater => {
+                                op_stack.push(stack_op);
+                                op_stack.push(next_op);
+                            }
+                            Less => {
+                                let lhs = arg_stack.pop().unwrap();
+                                let rhs = arg_stack.pop().unwrap();
```

Comment:
Not sure - was semi-copying [`Language.Haskell.Infix`](https://hackage.haskell.org/package/infix-0.1.1/docs/src/Language-Haskell-Infix.html), but will have to check. You are right to be concerned.